### PR TITLE
fixes #3405 - deleting large chunks of text caused a noticeable lag

### LINKED ISSUE #3405: Deleting large chunks of text causes a significant lag.
## Version
1.6.6

## Test Case
http://jsfiddle.net/2qa39t6g/2/

## Steps to reproduce
1. Click on the text and make sure the cursor is displayed.
2. Hit ctrl+a (cmd+a on macOS) to select all the text.
3. Hit backspace on your keyboard.

## Expected Behavior
The text should be removed immediately.

## Actual Behavior
There is a noticeable lag when removing large chunks of text.
This is caused by a time-expensive call to `this._splitTextIntoLines()` inside `fabric.IText.prototype._removeSingleCharAndStyle` (which is called for every character eligible for removal).

I will send a Pull Request shortly.

### PR DESCRIPTION
**Pull request by belfz**
This is a fix for [issue 3405](https://github.com/kangax/fabric.js/issues/3405).
It also fixes the `tests.mustache` file ([issue](https://github.com/kangax/fabric.js/issues/3404)), because I added unit tests and had to run them.

### CODE REVIEW

File: src/mixins/itext_behavior.mixin.js [hunk 1]
```diff
@@ -592,21 +592,24 @@
...
+        this._removeSingleCharAndStyle(start + 1, Math.abs(end - start) === 1);
         end--;
       }
       this.selectionStart = start;
       this.selectionEnd = start;
     },
 
-    _removeSingleCharAndStyle: function(index) {
+    _removeSingleCharAndStyle: function(index, shouldRewrapLines) {
+      shouldRewrapLines = typeof shouldRewrapLines === 'undefined' ? true : shouldRewrapLines;
```
Comment (belfz) [inline src/mixins/itext_behavior.mixin.js]: `shouldRewrapLines` is `true` by default.

File: tests.mustache [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <html>
 <head>
 <title>Test'em</title>
-<script src="/bower_components/qunit/qunit/qunit.js"></script>
+<script src="/node_modules/qunitjs/qunit/qunit.js"></script>
```
Comment (belfz) [inline tests.mustache]: Tests are actually fired now. However, some tests seem to fail (CLI report reports 1 failed test, while Chrome browser's result mentions 2 failed tests). They failed before and are not caused by my changes. Please see [this issue](https://github.com/kangax/fabric.js/issues/3404).

**Comment by asturur** [comment]
I have feelings this will break styles.
Can you try this code with style around text? i guess that when you delete chars that spans across lines, style will be broken.
Actually the old code was faster, this splitTextIntoLines was introduced by me both in the paste and delete code.

**Comment by belfz** [comment]
@asturur thanks - I am using this fix in the project I am working on and so far it seems to be working fine.

Characters are being removed one by one, in a loop. There is no need to fire the line-breaking algorithm after each character (`this._splitTextIntoLines()`) - it can be fired only once, at the very end of the loop.

Also, inside `_splitTextIntoLines` method, there is a call to `_generateStyleMap` method, which regenerates the styles map (based on the state of the text).

I have prepared a [test fiddle here](http://jsfiddle.net/LsLno527/3/). You can play around and verify that the fix doesn't break the styles (unless I am missing some case).

**Comment by asturur** [comment]
thanks. i ll give it a look. from my memories is the getcursorLocation, necessary for style shift, that needa that line refactor. textbox will be slower than itext i guess

**Comment by asturur** [comment]
Ok here is the problem.
Your logic works with Texbox but with not IText.

take this fiddle:
http://jsfiddle.net/LsLno527/8/

image

delete text like showed in picture and style will be messed up.
If you manage to make it work pretty i m more than happy to push it in.

**Comment by belfz** [comment]
Thank you @asturur. That's true, it doesn't seem to work correctly with the instances of `IText` (in the project I am working on I use `Textbox`). I might try to look for the solution to make it work in `IText`, but in the meantime let's put this pull request on hold.

**Comment by asturur** [comment]
a speed up could be gained trying to call the slow function for every deleted line.
In textbox with `\n` you will get the same problem.

I guess you can inspect the char deleted, check if it is a `\n` and in that case rewrap lines.

Not really sure, i do not remember everything but i think that before giving up on doing this operation every character, i tried all the first level solutions.

**Comment by belfz** [comment]
Closing this pull request, because the proposed solution breaks the styles in the text.

**Comment by asturur** [comment]
Can't you give a shot at detecting line deletion and call the `expensive` function just in that case?

**Comment by belfz** [comment]
@asturur I gave it a shot in [this updated fiddle](http://jsfiddle.net/LsLno527/10/)

I just changed that `if` statement in `_removeSingleCharAndStyle`:
```
if (isBeginningOfLine) {
    this._textLines = this._splitTextIntoLines();
}
```

Unfortunately, it still doesn't work correctly.

**Pull request closed without merging**